parser-utils/src/value_and_span.rs

```diff
@@ -0,0 +1,282 @@
+
+impl PartialOrd for SpanPosition {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for SpanPosition {
+    fn cmp(&self, other: &Self) -> Ordering {
+        (self.0).0.cmp(&(other.0).0)
+    }
+}
+
+/// An iterator specifically for iterating `edn::ValueAndSpan` instances in various ways.
+///
+/// Enumerating each iteration type allows us to have a single `combine::Stream` implementation
+/// yielding `ValueAndSpan` items, which allows us to yield uniform `combine::ParseError` types from
+/// disparate parsers.
+#[derive(Clone)]
+pub enum IntoIter {
+    Empty(std::iter::Empty<edn::ValueAndSpan>),
+    Atom(std::iter::Once<edn::ValueAndSpan>),
+    Vector(std::vec::IntoIter<edn::ValueAndSpan>),
+    List(std::collections::linked_list::IntoIter<edn::ValueAndSpan>),
+    /// Iterates via a single `flat_map` [k1, v1, k2, v2, ...].
+    Map(std::vec::IntoIter<edn::ValueAndSpan>),
+    // TODO: Support Set and Map more naturally.  This is significantly more work because the
+    // existing BTreeSet and BTreeMap iterators do not implement Clone, and implementing Clone for
+    // them is involved.  Since we don't really need to parse sets and maps at this time, this will
+    // do for now.
+}
+
+impl Iterator for IntoIter {
+    type Item = edn::ValueAndSpan;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match *self {
+            IntoIter::Empty(ref mut i) => i.next(),
+            IntoIter::Atom(ref mut i) => i.next(),
+            IntoIter::Vector(ref mut i) => i.next(),
+            IntoIter::List(ref mut i) => i.next(),
+            IntoIter::Map(ref mut i) => i.next(),
+        }
+    }
+}
+
+/// A single `combine::Stream` implementation iterating `edn::ValueAndSpan` instances.  Equivalent
+/// to `combine::IteratorStream` as produced by `combine::from_iter`, but specialized to
+/// `edn::ValueAndSpan`.
+#[derive(Clone)]
+pub struct Stream(IntoIter, SpanPosition);
+
+/// Things specific to parsing with `combine` and our `Stream` that need a trait to live outside of
+/// the `edn` crate.
+pub trait Item: Clone + PartialEq + Sized {
+    /// Position could be specialized to `SpanPosition`.
+    type Position: Clone + Ord + std::fmt::Display;
+
+    /// A slight generalization of `combine::Positioner` that allows to set the position based on
+    /// the `edn::ValueAndSpan` being iterated.
+    fn start(&self) -> Self::Position;
+    fn update_position(&self, &mut Self::Position);
+
+    fn into_child_stream_iter(self) -> IntoIter;
+    fn into_child_stream(self) -> Stream;
+    fn into_atom_stream_iter(self) -> IntoIter;
+    fn into_atom_stream(self) -> Stream;
+}
+
+impl Item for edn::ValueAndSpan {
+    type Position = SpanPosition;
+
+    fn start(&self) -> Self::Position {
+        SpanPosition(self.span.clone())
+    }
+
+    fn update_position(&self, position: &mut Self::Position) {
+        *position = SpanPosition(self.span.clone())
```

Comment:
Yeah, this is a weirdness arising from trying to agree with `combine`.  It's probably not worth agreeing with `combine`, but it's also not a big issue to do it.